druid/druid.lua

```diff
@@ -58,15 +58,65 @@ local _instances = {}
 
 local function get_druid_instances()
 	for i = #_instances, 1, -1 do
+		if _instances[i].set_pr:is_exist() then
+			_instances[i].set_pr:clear()
+		end
+		_instances[i].set_pr:subscribe(update_instances)
+		_instances[i].reset:subscribe(reset_instances)
 		if _instances[i]._deleted then
 			table.remove(_instances, i)
 		end
 	end
-
+	sort_instances()
 	return _instances
 end
 
 
+function update_priority(url, value, freeze_or_unfreeze)
+	for i = #_instances, 1, -1 do
+		if _instances[i].url == url then
+			_instances[i]._priority = value
+		else 
+			if freeze_or_unfreeze == "freeze" then
```

Comment:
> All code about freeze/unfreeze is redundant due the `BaseComponent:set_input_enabled` function. It's works on all components and don't pass the user input if it's disabled
This work is result of my student practice. And our tutor ask me to make switching between components in one click and reset input if there was interaction with another component , because it will be more comfortable for user. Switching between components which belong to one game object can be make with **BaseComponent:set_input_priority**, but it does not work for components which belong to different game objects. As I understand I can't set priority to game objects directly. I found out that call of **msg.post(, "acquire_input_focus")** will make game object the most prioritized, that gives ability to sort instances. But there was problem: if i select input component, it's priority and priority of instance must be lower then priority of another components and instances because I need switching in one click, if priority will be higher another will not be selected.  Selected input component will "eat" our click. Low priority cause problem which I solve with freeze/unfreeze. Components with keyboard input (buttons, back handler) will have higher priority than selected input, and they will read keyboard input, which they should not. Thats why I add ability to disable keyboard input. If I disable they will not be clickable. May be there could be another solution. May be global flag about selection of input component, but it also strange. May be You can recommend how block only keyboard input.


Здесь то же самое только на русском, так как я мог не совсем точно и корректно описать все:
Вся эта работа результат студенческой практики. Организация и куратор попросили ко всему прочему реализовать переключение и сброс Input в один клик, аргументируя тем, что переключение минимум через два неудобно для пользователя. И если сделать переключение между элементами внутри одного game object не является проблемой за счет **BaseComponent:set_input_priority**, то для того, чтобы это провернуть с элементами из разных, нужно регулировать приоритет instances и game objectes, единственное, что я смог придумать, это использовать **msg.post(, "acquire_input_focus")**, что делает game object наиболее приоритетным в плане ввода, как я понял. Для того, чтобы нажатия обрабатывались в один клик, приоритет выбранного input элемента должен быть ниже, чем остальные элементы, для элементов из разных game objects необходима чтобы instance имел меньший приоритет относительно остальных иначе могут происходить ситуации, что у нас выбрано 2+ input элемента или выбор в 2 клика, что меня попросили исправить. Но все это вызывает проблему, что при вводе в выбранный input, другие элементы (button, back handler) имея больший приоритет съедают ввод, для этого я и добавил freeze/unfreeze. Если Вы подскажите, как можно более лаконично и корректнее решить этот вопрос, то это будет отлично. Поскольку мне первое, что приходит в голову создать глобальный флаг о том, что у меня ввод в input, что так себе идея.

Reply:
Давай на русском)

Там в input компоненте я начал добавлять для этой фичи - NO_CONSUME_INPUT_WHILE_SELECTED, но надо проверить что и как работает, т.к. помню что не все оч хорошо с ним

В любом случае, чтобы дальше с PR поработать, решение с фриз и прочими нужно убрать, а все изменения для input - сделать только в input. По идее самого кода так много не будет

Когда такой объем ПР и кажется, что изменения не оч связанны - сложней разбираться

Глобальный флаг - плохо, т.к. сложно контролируется и могут быть коллизии разные

Можешь посмотреть на NO_CONSUME_INPUT_WHILE_SELECTED и глянуть, похоже ли на это и может в ту сторону покопаем? что думаешь?

Ну и остальные комменты актуальные к ПР

Reply:
Я посмотрел NO_CONSUME_INPUT_WHILE_SELECTED, и понял, что то, что я хотел, можно было сделать парой строчек.
В итоге вышло, что NO_CONSUME_INPUT_WHILE_SELECTED можно и не включать, однако я не знаю, насколько найденное решение нормальное.
В Input.on_input, где мы обрабатываем actions я добавил:		
_if action_id == const.ACTION_TOUCH or action_id == const.ACTION_MULTITOUCH then
		return **false**
end_
Тем самым действия мышью не должны захватываться Input.

Но перед этим я включил NO_CONSUME_INPUT_WHILE_SELECTED и попытался добиться желаемого эффекта. При наивысшем приоритете, actions, обработка которых в условии имела часть "action.released", срабатывает ок, но если там что-то иное, как у backspace: action.pressed or action.repeated, то при наличии, например, backhandler, который срабатывает на backspace, происходит перехват backhandler'ом нажатой клавиши, то же самое, если на экране будут существовать кнопки, имеющие key_trigger на пробел (ну или клавишу символьную).
Попытка поиграться с if не дала особых результатов. Стало понятно, что нужно как-то сепарировать ввод мышки от клавиатуры. И самым простым показалось обрабатывать ACTION_TOUCH и т.п. и возвращать false.

-------------------------------------------------------------------------------------------------------------------------------------

Проверял я в начале на примере, который внутри библиотеки, но потом решил попробовать это же только создав другой. Там не использовалась collection factory, а просто создано 2 game object'a, каждый из которых со своим gui. И тут началась та же проблема: кнопка из другого go готова "съесть" любое нажатие backspace, enter, пробел, в зависимости от того, какой я key_trigger поставлю. А внутри go кнопка работала нормально. Вероятно в выполненной до этого работе я что-то поломал, т.к. у меня и в библиотечном примере back_handler съедал backspace, это, в принципе, и стало причиной добавления freeze/unfreeze. Но между go захват ввода все равно возникает.

Решение рабочее для примера, вроде как, найдено. Но в общем я еще подумаю, что можно придумать, чтобы между двумя независимыми go не было проблем. Может у Вас есть какие-то идеи почему это так и как теоретически это можно поправить?